### ISSUE TITLE
Text to Speech (TTS)

### ISSUE THREAD

**Issue by Sawasdee** [koreader/koreader#545]
Is there anyway to add TTS to the latest koreader, please?   I've seen that in the previous version (kdfviewer) has this ability but not in the latest version.   I tried to learn the TTS code in the previous version (dialog.lua), but it was written in the entirely different way the latest version written.   So I'm really don't know how to modified or use the code.   Any help or suggestion, please?   Thanks a lot.

**Comment by Korb** [comment]
Neither in the [README.md](https://github.com/koreader/koreader#readme) file, nor in the [User Guide](https://koreader.rocks/koreader-user-guide.pdf), nor on [the application site](https://www.google.com/search?q=android+tts+site%3Akoreader.rocks), I found a mention of TTS. Do I understand correctly that this application, at least its version for Google Android, does not have the ability to read open text documents by voice?

**Comment by poire-z** [comment]
> this application, at least its version for Google Android, does not have the ability to read open text documents by voice?

You understand correctly.
There is no code in the app to generate speech and neither do anything related to audio.
See also #9548 and #7975.

**Comment by foss-** [comment]
> Do I understand correctly that this application, at least its version for Google Android, does not have the ability to read open text documents by voice?

Been asking myself the same for a long time.

https://github.com/koreader/koreader/issues/9548 seems to be the correct feature request. While I understand there currently is no interest in implementing a solution, I am not sure I understand why this valid feature request would be closed.

**Comment by NiLuJe** [comment]
> I am not sure I understand why this valid feature request would be closed.

Because it's out of scope. For technical reasons, which then snowballs into a human resources reason, as it would take an enormous amount of work.

**Comment by ryanwwest** [comment]
@NiLuJe would it be appropriate to build a plugin for this, that for example, only works on android to start? I see some of the plugins are only for specific devices, but I don't really know how much work it might take to integrate it with KOReader since it has a totally customized UI.

**Comment by NiLuJe** [comment]
I'm not sure if a plugin would have the required access necessary to even attempt to implement this. But, yes, from a design standpoint, that's probably better. It'would most likely forever be stuck in contrib land because of the aforemenetion reasons, though.

**Comment by ryanwwest** [comment]
Other ereader apps supporting TTS like Moon+ Reader and Librera Reader seem to use an Android TTS service where different TTS-specific apps like [Speech Services by Google](https://play.google.com/store/apps/details?id=com.google.android.tts&hl=en_US&gl=US) (other open-source ones are not all that good yet) register as possible Android TTS services that can be selected in Android settings by Moon/Librera. Thus, my inclination is that the plugin would need access to a few related Android settings for piping the text to be read to that API so the other app can dictate them (while not leaving the context of koreader's UI). 

The KOReader setting Tools -> Plugin management seems to have everything enabled by default - I'm not sure if that's for the same type of plugin that I'm talking about (but maybe I'm wrong and this is it for plugins). Regardless, maybe a patch would be better to allow for access like this, if Android SDK access can even be changed/used this way.

**Comment by NiLuJe** [comment]
We're also not actually an Android "app", but more like a game engine, so I'm not even sure anything like can that could even technically apply.

**Comment by Frenzie** [comment]
> so I'm not even sure anything like can that could even technically apply.

It might be relatively "straightforward" to provide an "accessibility server". Where by straightforward I mean it's a lot of effort, but much of the secondary part of it we've already done in FocusManager, adding descriptions (mainly for debug log) that could be reused, etc. The obstacle would be reimplementing whatever macOS and/or Qt/Gtk do but we wouldn't be starting from scratch.

Android throws a major wrench in the works though. That'd be significantly harder than a relatively sane OS like Win/Mac/normal Linux, if it's even possible to fake the intent system that way at all.

**Comment by pazos** [comment]
> Regardless, maybe a patch would be better to allow for access like this, if Android SDK access can even be changed/used this way.

Android SDK cannot be used by patches or plugins without glue code written first into https://github.com/koreader/android-luajit-launcher

> We're also not actually an Android "app", but more like a game engine, so I'm not even sure anything like can that could even technically apply.

That's the problem.

The `MainActivity` is indeed using the NativeActivity framework nobody else is using :p

android-luajit-launcher is mostly glue code to use sdk only methods (not available via NativeActivity framework).
While the basic implementation of NativeActivity is based on function callbacks and not very flexible we use android_app_glue to gain some flexibility but we screw its multithread awareness by forwarding everything to LuaJIT (we keep or own fork of android_app_glue just to be able to pass BroadcastReceiver events to lua land) 

Everything works :tm: just because LuaJIT is fast enough to handle the entire android app state, react to input events and draw to canvas in time.

Also the frontend/base has the same kind of relationship with LuaJIT and everything that might block, such as network things, is handled by LuaJIT. That's not a problem outside android. Android expects all jobs to be done outside the UiThread, and we should use the UiThread just to call methods that update the view hierarchy and that's **NOT** happening here.

As you can see both approaches are mostly incompatible and yet we're here thanks to Mike Pall and a bunch of hacks :)

But it is doable on principle. Just make TTS a service and use the MainActivity to start/stop the TTS service engine.

Doable doesn't mean a good idea, easy, fun or sane by any quality of life metrics.

> Android throws a major wrench in the works though. That'd be significantly harder than a relatively sane OS like Win/Mac/normal Linux, if it's even possible to fake the intent system that way at all.

I would start the development outside android. I have no idea of the work involved in crengine & mupdf glue to get the effect we call TTS (highlight chunks of text at a given speed and get the text ready to be sent somewhere).

If somebody attempts to do something like that, as a patch or an user plugin, **do not start with the android part**. First do the fun stuff. The android specific bits might be possible with even more hacks on top of our hacks :see_no_evil: 

Just my two cents.

**Comment by ryanwwest** [comment]
> We're also not actually an Android "app", but more like a game engine

Interesting - I was wondering about the possibility of building other e-ink-friendly apps in this since it has such a unique UI. I get if it's so far separated from regular Android apps than this type of integration would be tough (not to mention supporting so many other devices). I was scratching my head wondering how a repo with 95%+ Lua compiles down to Java/Kotlin... Thanks for your input as well, @pazos. It's helpful to see the large scope that would be required for something like this.

Perhaps a simpler alternative that gets you a lot of the way there on Android is to just use another app that supports TTS when you want to use it. But there is currently a little friction with opening the other app, navigating to the current file path being read in KOReader, and opening it / starting TTS in that app. It would make it a lot easier if one could quickly jump from KOReader directly to that other app and open the same book in it, which could probably be accomplished using [Android Sharesheets](https://developer.android.com/training/sharing/send) - what most apps rely on when a user clicks the 'Share' button. Android maintains a list of which apps can handle which data types (like epub file, html file etc.) and will let the user select them.

I wonder if adding something like this would be possible, since I think the only thing KOReader would really need to provide is the absolute filepath of the currently opened file (or currently selected file if in the file browser)? I've noticed that when a PDF reflow bug crashes the app on 2023.04 Android version and the log is shown, clicking the 'Save Report' button brings up the sharesheet (to save in another app) so it seems this particular Android integration is already implemented in the app. Not even the current page or other metadata would be needed and I doubt that is cross-app-friendly anyway.

Note I'm not talking about sharing a file _to_ KOReader - I've read discussions about the complexity of copying the file over, if it's on an SD card etc. I'm only talking about some sort of 'share' button _within_ KOReader to an _external_ app. For TTS, for viewing an HTML more natively in a browser app, etc. If this already exists and I missed it, please let me know :)

**Comment by pazos** [comment]
Yup, we use intents extensively and a bit of sharesheets too.

To:

Let you import files from android filepicker
Share chunks of highlighted text to another app
Share chunks of text to specific 3rd party dictionary apps
Run the installer when an update is downloaded and the user press the "install" button
Send you to wifi settings, to battery saving settings ...
...

But you can not pass file paths outside the app boundary (I think it happened on API24?). You need to wrap them up on a content scheme uri (like in https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/ApkUpdater.kt#L63-L67) and trust the other apps they're able to handle them.

The same happens when the source of the file is another app and we receive it. We just use another hack to guess the path, but it has some issues, as you can see in https://github.com/koreader/koreader/issues/9787

**Comment by ryanwwest** [comment]
I see. Would it be difficult to wrap the filepath in a content scene uri and offer this, accepting the possibility that some of those issues arise? I didn't fully grasp all the discussion in the links you mentioned but it seems like many of the problems originate from trying to share a file into Koreader, not the other way around. Granted I know very little about this architecture.

**Comment by pazos** [comment]
> I see. Would it be difficult to wrap the filepath in a content scene uri and offer this, accepting the possibility that some of those issues arise? I didn't fully grasp all the discussion in the links you mentioned but it seems like many of the problems originate from trying to share a file into Koreader, not the other way around. Granted I know very little about this architecture.

Nope, that's easy :)

See the slightly outdated https://github.com/koreader/koreader/pull/5762 and https://github.com/koreader/android-luajit-launcher/pull/218/files.

**Comment by ryanwwest** [comment]
I see, so it's at least partially implemented (last link) but waiting to officially merge until more people request it? If that's the case, I'll request it. Lol. I think it could make TTS and other cross-app use cases a lot easier. Perhaps, in addition to in the file browser (as the first link suggested), as a context menu item within the Hamburger menu, maybe near Book Status? As then you don't have to find and reselect the document in the browser in order to share it.

f both are too much work, even just the Hamburger icon item alone would be enough. I'm not sure if you'd hide the option in other builds like Linux that don't have a corollary though (unless you can copy the book path into the clipboard instead).

**Comment by antrmn** [comment]
This would be such a nice feature to have on KOReader. 

I wonder if the KOReader APIs already provide a way to play audio. One could glue up a small TTS program like [piper](https://github.com/rhasspy/piper) or spd-say frontend to play out the current page (or selection).

**Comment by ryanwwest** [comment]
I wish this too, but poire-z said:

> There is no code in the app to generate speech and neither do anything related to audio.

I don't know if it would be that hard to add, though.

**Comment by poire-z** [comment]
>  If the KOReader APIs were to provide a way 

We don't expose any API.
Or... we do expose all of KOReader to plugins and user-patches :)

So, I guess one could create a plugin with a `:onPageUpdate()` that would get called on each new page displayed: it could then extract the whole page text (cf `ReaderView:getCurrentPageLineWordCounts()`, which gets it to count lines and words), massage it to make sentences, and feed that (your challenge, not ours: we don't want to get involved, not our scope, too variable by device type) to some external program/process that will do the TTS. (And `:onClose()` to cancel things.)

**Comment by C-Loftus** [comment]
In my opinion, I think it may make sense to write a plugin for a Lua speech dispatcher client inside Koreader. It would mainly just work for those using a Linux-based ereader or Desktop, but it is fairly flexible as speech dispatcher works both on device or over the network and doesn't require API keys like a proprietary TTS service would.  Speech dispatcher handles all the essentials for a speaking application like queueing, canceling etc. and recent versions include [piper](https://github.com/rhasspy/piper) out of the box which is quite natural sounding. 

I wanted to write some thoughts in case others had opinions.  To implement this I would:

- Write a lua speech dispatcher client
    - I am less familiar with Lua / Rust interop but there's a chance you could also just directly use the [Rust speech dispatcher client library](https://github.com/odilia-app/ssip-client-async)
- Watch for `:onPageUpdate()`
- Use `ReaderView:getCurrentPageLineWordCounts()` to get all text on screen
- Split by a delimeter like `.` and send each sentence to the speech dispatcher daemon
    - You could also just send the entire page at once; doesn't make much difference afaik
- speech dispatcher internally has a queue so there is no issue if multiple concurrent requests are sent
- if the user flips the page while content is being spoken, send a message to the speech dispatcher daemon to cancel all speech in the queue 
- if all the messages are spoken and speech dispatcher returns success you can auto advance the page 
    - I presume koreader has an API for this but am unsure as I am new to plugin development

If anyone thinks differently or would want to collaborate on a plugin feel free to message me. I wrote this here since it seems to be the canonical issue on TTS.

**Comment by poire-z** [comment]
Your logic looks fine.

> Write a lua speech dispatcher client
>     * I am less familiar with Lua / Rust interop but there's a chance you could also just directly use the [Rust speech dispatcher client library](https://github.com/odilia-app/ssip-client-async)

We'd probably don't want to have that bit integrated in KOReader - but you could probably communicate with the text-to-speech converter over a fifo or a unix socket, which is what your link seems to allow.
You should probably try to see if something like that is possible on Android (I guess most users asking for the features would like to listen to their book on the go, and switch to reading on the no-go) ; if there would be a common solution, the better.

**Comment by ryanwwest** [comment]
+1 for android and desktop Linux compatibility.

**Comment by C-Loftus** [comment]
> We'd probably don't want to have that bit integrated in KOReader - but you could probably communicate with the text-to-speech converter over a fifo or a unix socket, which is what your link seems to allow.

Yup, I think we are on the same page; speech dispatcher communicates over a socket and does all the text to speech conversion serverside. The client for speech dispatcher in this scenario is similar to an http client on localhost and thus is always doing the CPU bound work outside of KOreader (either on the same device or a remote). i.e. an old Kindle device could just send the payload to a Desktop Linux device on the same network without needing to do anything on device necessarily

From what I have read in this thread I am not sure how one would support Android in a straightforward way. Even if you manage to get in an environment where you can hook into the Android host running KOReader, the Android speech server API, [TextToSpeech](https://developer.android.com/reference/android/speech/tts/TextToSpeech), which is the Android equivalent of Speech Dispatcher to my understanding, does not support communication via a socket so you would need ffi with Java I think? But if that was done then someone would need to write a generic wrapper over both that and speech dispatcher. This final step is not particularly hard since they are a very similar API. At least for me I don't have an Android device so this isn't in my personal plan at the moment, but would be happy to offer input on anything re interoperability.  

So in short, I think I agree with the comment in the thread above

> If somebody attempts to do something like that, as a patch or an user plugin, do not start with the android part. First do the fun stuff. The android specific bits might be possible with even more hacks on top of our hacks 🙈

I will probably tinker on this in upcoming weekends and report back if I have any success. May be a while since the lua client will take some time.

**Comment by antrmn** [comment]
What about Libspiel instead of Speech dispatcher? 
https://project-spiel.org/

**Comment by Frenzie** [comment]
> so you would need ffi with Java I think?

Presumably you'd write almost everything in Java/Kotlin and then basically just pass the text and maybe some options with FFI/JNI, yes.

Not unlike the clipboard:
https://github.com/koreader/koreader/blob/8a7e077cdd1223b91837d807a6750e7b4d2230b6/frontend/device/android/device.lua#L235-L243

**Comment by hius07** [comment]
> Use `ReaderView:getCurrentPageLineWordCounts()` to get all text on screen

More appropriate text extraction in:
https://github.com/koreader/koreader/blob/8a7e077cdd1223b91837d807a6750e7b4d2230b6/frontend/apps/reader/modules/readerhighlight.lua#L2012
Can be moved to a separate method `ReaderView:getCurrentPageText()`.

**Comment by pazos** [comment]
@C-Loftus:

> From what I have read in this thread I am not sure how one would support Android

Ideally we wouldn't need too. This is android. There're a lot of TTS apps ;)

We would need a bunch of new permissions to handle calls, notifications, bluetooth devices? et all. That would be a little bit messy.

But if the "general support" works on every other device I would say it is a huge success :)

**Comment by shuvashish76** [comment]
TTS plugin https://github.com/norude/TTS.koplugin
Didn't test.

**Comment by auspiciouslink** [comment]
If I'm understanding the setup correctly, might it even be possible to connect back to the home network over head/tailscale for remote access?

**Comment by Artem89233** [comment]
Hi everyone, how's tts working?

**Comment by amadib** [comment]
> [@C-Loftus](https://github.com/C-Loftus):
> 
> > From what I have read in this thread I am not sure how one would support Android
> 
> Ideally we wouldn't need too. This is android. There're a lot of TTS apps ;)
> 
> We would need a bunch of new permissions to handle calls, notifications, bluetooth devices? et all. That would be a little bit messy.
> 
> But if the "general support" works on every other device I would say it is a huge success :)

I understand that there's a lot of Android TTS apps but running a server seems so convoluted and more of a workaround than proper  solution. Don't get me wrong I'd love to see this but after reading through the conversations it makes sense what the devs are saying to have this as a plugin and separated from the core functionality to keep things clean. 

That being said I am confused what would prevent support with the native [Assistive Readers](https://www.amazon.com/-/en/gp/help/customer/display.html?nodeId=TdHCSXXqsdhaVKt3Tc) that are natively available for kindles?

**Comment by ryanwwest** [comment]
https://github.com/norude/TTS.koplugin is a cool proof of concept on Android - you can use Termux to run the edge.tts server script that uses Microsoft's online TTS with no signup and little configuration. Though it's buggy and for edge.tts must be online - there is also a local-run Piper TTS option that works with this, but I haven't gotten it to work yet.

I wonder if a third backend for this could be be just selecting an Android TTS app as discussed. The UI in this plugin looks pretty nice so far, despite the bugs.

**Comment by poire-z** [comment]
Another plugin being worked on, quite impressive with the platforms it supports and its documentation: https://github.com/stradichenko/audiobook.koplugin
